src/System.Memory/src/System/Span.cs

```diff
@@ -0,0 +1,400 @@
+        }
+
+        /// <summary>
+        /// Forms a slice out of the given span, beginning at 'start', of given length
+        /// </summary>
+        /// <param name="start">The index at which to begin this slice.</param>
+        /// <param name="length">The desired length for the slice (exclusive).</param>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified <paramref name="start"/> or end index is not in range (&lt;0 or &gt;=Length).
+        /// </exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Span<T> Slice(int start, int length)
+        {
+            if ((uint)start > (uint)_length || (uint)length > (uint)(_length - start))
+                throw ThrowHelper.CreateThrowArgumentOutOfRangeException(ExceptionArgument.start);
+
+            IntPtr newOffset = _byteOffset.Add<T>(start);
+            return new Span<T>(_pinnable, newOffset, length);
+        }
+
+        /// <summary>
+        /// Copies the contents of this span into a new array.  This heap
+        /// allocates, so should generally be avoided, however it is sometimes
+        /// necessary to bridge the gap with APIs written in terms of arrays.
+        /// </summary>
+        public T[] ToArray()
+        {
+            if (_length == 0)
+                return SpanHelpers.PerTypeValues<T>.EmptyArray;
+
+            T[] result = new T[_length];
+            CopyTo(result);
+            return result;
+        }
+
+        /// <summary>
+        /// Returns a 0-length span whose base is the null pointer.
+        /// </summary>
+        public static readonly Span<T> Empty = default(Span<T>);
+
+        /// <summary>
+        /// Returns a reference to the 0th element of the Span. If the Span is empty, returns a reference to the location where the 0th element
+        /// would have been stored. Such a reference can be used for pinning but must never be dereferenced.
+        /// </summary>
+        public ref T DangerousGetPinnableReference
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                if (_pinnable == null)
+                    unsafe { return ref Unsafe.AsRef<T>(_byteOffset.ToPointer()); }
+                else
+                    return ref Unsafe.AddByteOffset<T>(ref _pinnable.Data, _byteOffset);
+            }
+        }
+
+        //
+        // If the Span was constructed from an object,
+        //
+        //   _pinnable   = that object (unsafe-casted to a Pinnable<T>)
+        //   _byteOffset = offset in bytes from "ref _pinnable.Data" to "ref span[0]"
+        //
+        // If the Span was constructed from a native pointer,
+        //
+        //   _pinnable   = null
+        //   _byteOffset = the pointer
+        //
+        private readonly Pinnable<T> _pinnable; 
+        private readonly IntPtr _byteOffset;
+        private readonly int _length;
```

Comment:
Please move these to the top of the type. I know it's not your desired style, but it's how 99.9% of the rest of corefx is, with the stragglers being things we should fix as well.